server/src/graql/reasoner/rule/RuleUtils.java

```diff
@@ -128,11 +133,41 @@
 
     /**
      * @param rules set of rules of interest forming a rule subgraph
-     * @return true if the rule subgraph formed from provided rules contains loops
+     * @return true if the rule subgraph formed from provided rules contains loops AND corresponding relation instances also contain loops
      */
-    public static boolean subGraphIsCyclical(Set<InferenceRule> rules){
-        return !new TarjanSCC<>(persistedTypeSubGraph(rules))
-                .getCycles().isEmpty();
+    public static boolean subGraphIsCyclical(Set<InferenceRule> rules, TransactionOLTP tx){
+        HashMultimap<Type, Type> typeSubGraph = persistedTypeSubGraph(rules);
+        TarjanSCC<Type> typeSCC = new TarjanSCC<>(typeSubGraph);
+        List<Set<Type>> typeCycles = typeSCC.getCycles();
+        HashMultimap<Type, Type> typeTCinverse = HashMultimap.create();
+
+        typeSCC.successorMap().entries().forEach(e -> typeTCinverse.put(e.getValue(), e.getKey()));
+
+        return typeCycles.stream().anyMatch(typeSet -> {
+            Set<ReasonerAtomicQuery> queries = typeSet.stream()
+                    .flatMap(type -> typeTCinverse.get(type).stream())
+                    .flatMap(type -> tx.queryCache().getFamily(type).stream())
+                    .map(Equivalence.Wrapper::get)
+                    .filter(q -> tx.queryCache().getParents(q).isEmpty())
+                    .filter(q -> q.getAtom().isRelation() || q.getAtom().isResource())
+                    .collect(toSet());
+            if (!queries.stream().allMatch(q -> tx.queryCache().isDBComplete(q))) return true;
+
+            HashMultimap<Concept, Concept> conceptMap = HashMultimap.create();
+            return queries.stream().anyMatch(q -> {
+                RelationAtom relationAtom = q.getAtom().toRelationAtom();
+                Set<Pair<Variable, Variable>> varPairs = relationAtom.varDirectionality();
+                IndexedAnswerSet answers = tx.queryCache().getEntry(q).cachedElement();
+                return answers.stream().anyMatch(ans ->
+                        varPairs.stream().anyMatch(p -> {
+                            Concept from = ans.get(p.getKey());
+                            Concept to = ans.get(p.getValue());
+                            if (conceptMap.get(to).contains(from)) return true;
+                            conceptMap.put(from, to);
+                            return false;
+                        }));
+            });
+        });
```

Comment:
maybe in the future the above method can be made a bit more readable? even though I understand it might be complex